igvc_perception/src/train_eval/evaluate.py

```diff
@@ -0,0 +1,130 @@
+                 transforms.ToTensor(),
+            ])
+if __name__ == '__main__':
+    args = parser.parse_args()
+    img_dir = args.img_dir
+    im_size = args.im_size
+    anno_dir = args.anno_dir
+
+    model = UNet([3,im_size[0],im_size[1]], 3)
+    model.load_state_dict(torch.load(args.load_model))
+
+    img_names = np.sort(os.listdir(img_dir))
+    if anno_dir:
+        anno_names = np.sort(os.listdir(anno_dir))
+        assert len(img_names) == len(anno_names)
+
+    model.cuda()
+    model.eval()
+
+    # Calculate statistics.
+    inference_times = []
+    avg_precisions = {}
+    avg_recalls = {}
+    avg_accuracy = 0.0
+
+    for i in range(len(img_names)):
+        img_name = img_names[i]
+        print('Evaluating %s' % img_name)
+        img = cv2.imread(os.path.join(img_dir,img_name))
+        img = cv2.resize(img, (im_size[0],im_size[1]))
+
+        # Get inference time.
+        start_t = timeit.default_timer()
+
+        img_to_tensor = Image.fromarray(img)
+        img_tensor = transform(img_to_tensor)
+        img_tensor = Variable(img_tensor.unsqueeze(0)).cuda()
+
+        #img_tensor = Variable(img_tensor.unsqueeze(0))
+
+        output = model(img_tensor)
+        output_data = output.cpu().data.numpy()[0][0]
+
+        end_t = timeit.default_timer()
+        inference_times.append(end_t-start_t)
+
+        if anno_dir:
+            anno_name = anno_names[i]
+            anno = cv2.imread(os.path.join(anno_dir, anno_name))
+            anno[anno!=0] = 255
+            if anno.shape[2] == 3:
+                anno = anno[:,:,2]
+                anno = cv2.resize(anno, (im_size[0],im_size[1]))
+
+            # Calculate accuracy, where any value above 0.5 is true, below is false.
+            acc = (np.sum(output_data[anno!=0] > 0.5) + np.sum(output_data[anno==0] < 0.5)) / float(im_size[0]*im_size[1])
+            avg_accuracy += acc
+
+            # Calculate precision and recall for different thresholds.
+            for thresh in np.arange(0,1,0.05):
+                recall = np.sum(output_data[anno!=0] > thresh) / np.sum(anno!=0)
+                precision = np.sum(anno[output_data > thresh] != 0) / np.sum(output_data > thresh)
+
+                if not thresh in avg_precisions:
+                    avg_precisions[thresh] = 0.0
+                    avg_recalls[thresh] = 0.0
+                avg_recalls[thresh] += recall
+                avg_precisions[thresh] += precision
+
+        if args.vis:
+            cv2.imshow('input: ', img_tensor.cpu().data.numpy()[0][0])
+            cv2.imshow('output: ', output_data)
+
+            if anno_dir:
+                cv2.imshow('gt: ', anno)
+
+            k = cv2.waitKey(50)
+
+    if anno_dir:
+        avg_accuracy /= len(img_names)
+        recalls = []
+        precisions = []
+        keys = avg_recalls.keys()
+        keys.sort()
+        for key in keys:
+            avg_recalls[key] /= len(img_names)
+            avg_precisions[key] /= len(img_names)
```

Comment:
`avg_recalls` and `avg_precisions` could be quite easily fused to calculate the F1 score as the harmonic mean of precision and recall. [source](https://en.wikipedia.org/wiki/F1_score#Definition)

Reply:
added